### PR TITLE
RFD 56: SQL Backend

### PR DESCRIPTION
**Pull request by jimbishopp**
This PR adds an RFD describing requirements for the implementation of a PostgreSQL and CockroachDB backend supporting both self-hosted and cloud-hosted configurations.

### CODE REVIEW

**Comment by jimbishopp** [comment]
@r0mant There is not much here, but wanted to get some feedback. I have a lot of implementation notes, but removed them from the RFD. LMK what you'd like to see.

File: rfd/0056-cockroach-backend.md [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+for Teleport customers by allowing them to use existing infrastructure.
+
+## Scope
+
+This RFD focuses on implementing a backend for a self-hosted CockroachDB cluster.
+
+Requirements:
+
+- Prioritize self-hosted CockroachDB.
+- Implementation should support CockroachDB node failures by automatically
```
Comment (r0mant) [inline rfd/0056-cockroach-backend.md:25]: Can the driver handle this automatically or will we need to implement this ourselves?

File: rfd/0056-cockroach-backend.md [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+  reconnecting to another node when a connected node becomes unavailable.
+- Design should consider extensibility to support other SQL databases
```
Comment (r0mant) [inline rfd/0056-cockroach-backend.md:27]: I would maybe rephrase this a little bit to indicate that Postgres support (incl. cloud-hosted) is also in scope and other SQL databases are out of scope but the design would allow for extensibility.

File: rfd/0056-cockroach-backend.md [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+- Prioritize self-hosted CockroachDB.
+- Implementation should support CockroachDB node failures by automatically
+  reconnecting to another node when a connected node becomes unavailable.
+- Design should consider extensibility to support other SQL databases 
+  such as PostgreSQL and MySQL.
+- Design should utilize a dialer/connector abstraction when connecting to
+  the database to allow the implementation to be extended to support SQL
+  databases hosted by a cloud provider.
+
+## Authentication
```
Comment (r0mant) [inline rfd/0056-cockroach-backend.md:33]: Add a section about cloud authentication too? I.e. using IAM auth for RDS, Cloud SQL, etc.

File: rfd/0056-cockroach-backend.md [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+The backend will require mutual authentication when connecting to the CockroachDB 
+cluster. This requires users to generate client certificates and configure
```
Comment (r0mant) [inline rfd/0056-cockroach-backend.md:36]: I'm thinking, instead of using Cockroach's client certificates why not use the same approach as with Database Access and use Teleport-generated certificates (Cockroach database will need to be configured with Teleport's CA)? Then you'll be able to configure your Cockroach server same way for both database access and as a backend. We will need just an extra field like `user` in the configuration to indicate which user Teleport should use when connecting. What do you think?

File: rfd/0056-cockroach-backend.md [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+Teleport users enable the CockroachDB Backend by setting the storage
+type in the Teleport configuration file to `cockroach`, setting a list 
+of accessible CockroachDB nodes, and adding paths to the location of 
+CockroachDB client certificates.
+
+```yaml
+teleport:
+  storage:
+    type: cockroach
```
Comment (r0mant) [inline rfd/0056-cockroach-backend.md:52]: Nit: I'd name this `cockroachdb` to be consistent with the database access protocol name we already use (or maybe we can support both).

Also, what about Postgres? Is it going to be `type: postgres`?

File: rfd/0056-cockroach-backend.md (hunk 4 above)
Comment (stevenGravy) [reply rfd/0056-cockroach-backend.md:36]: Isn't there a bit of chicken/egg.  HOw do you get a cert for your backend if it hasn't started to store somewhere?

File: rfd/0056-cockroach-backend.md (hunk 3 above)
Comment (jimbishopp) [reply rfd/0056-cockroach-backend.md:33]: Will do. I wasn't focused on cloud initially.

File: rfd/0056-cockroach-backend.md (hunk 5 above)
Comment (jimbishopp) [reply rfd/0056-cockroach-backend.md:52]: I wanted to name it roach :) I'll keep it consistent and add postgres.

File: rfd/0056-cockroach-backend.md (hunk 4 above)
Comment (r0mant) [reply rfd/0056-cockroach-backend.md:36]: Good point 😄 Let's keep the current approach with Cockroach-signed certs for now, we can revisit later if needed.

File: rfd/0056-cockroach-backend.md (hunk 1 above)
Comment (jimbishopp) [reply rfd/0056-cockroach-backend.md]: I'd like to hold off on this for now because I'm not sure of the value and am concerned about complexity. Connecting to multiple nodes assumes each node is consistent. This RFD focuses on postgres and cockroachdb, and supporting multiple nodes is really only a cockroachdb feature. We'd have to deal with scenarios like detecting whether the fallback node is in the same state as the original node. Considering most cloud providers have their own solutions for HA that allow clients to specify a single endpoint, this additional work would be of questionable value.

File: rfd/0056-sql-backend.md [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+    # Database name that Teleport will use to store its data.
+    # The database must not be shared.
+    database: "teleport"
+
+    # Database user that Teleport will use to connect to the database. The user 
+    # must be granted ownership of the database and the ability to login.
+    user: "teleport"
+
+    # TLS validation and mutual authentication.
+    tls:
```
Comment (jimbishopp) [inline rfd/0056-sql-backend.md:83]: The storage section doesn't currently support nesting.

The YAML lib decodes this into a `map[interface{}]interface{}` which then cannot be marshalled into JSON which is required when converting `backend.Params` to a `Config` struct (error is `json: unsupported type: map[interface {}]interface {}`). I'm investigating whether the storage section can be upgraded to support nesting.

Comment (jimbishopp) [reply rfd/0056-sql-backend.md:83]: Our YAML lib has changed this behavior in `v3` to decode to `map[string]interface{}` when all keys are strings. [This issue](https://github.com/go-yaml/yaml/issues/384) describes the change.

File: rfd/0056-sql-backend.md [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+SQL Backend by configuring the storage section in the Teleport configuration
+file. Setting the storage type to either `postgres` or `cockroachdb` enables the
+SQL Backend. Additional configurations may apply depending on whether the
+configuration is for self-hosted or cloud-hosted environments.
+
+```yaml
+teleport:
+  storage:
+    # Type of storage backend (postgres or cockroachdb).
+    type: postgres
```
Comment (r0mant) [inline rfd/0056-sql-backend.md:69]: Could you add a section somewhere in the RFD explaining what the difference will be between setting this to `postgres` vs `cockroachdb`? Will those basically be identical or will there be any difference in Teleport's behavior?

File: rfd/0056-sql-backend.md [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+    # Database connection endpoint.
+    uri: "postgres.example.com:5432"
+
+    # Database name that Teleport will use to store its data.
+    # The database must not be shared.
+    database: "teleport"
+
+    # Database user that Teleport will use to connect to the database. The user 
+    # must be granted ownership of the database and the ability to login.
+    user: "teleport"
```
Comment (r0mant) [inline rfd/0056-sql-backend.md:80]: Is this field really needed though? Both Postgres and Cockroach map the database user from the CN field in the client certificate.

File: rfd/0056-sql-backend.md [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+
+Mutual TLS authentication is optionally supported for GCP Cloud SQL.
+Paths to a client certificate and key must be added to the Teleport
+configuration file to enable mTLS.
+
+
+## UX
+
+Teleport users must first configure the instance and database where Teleport will 
+store its data. A new database instance and user must be created. The new user
```
Comment (r0mant) [inline rfd/0056-sql-backend.md:55]: Can Teleport create the database itself (given enough permissions)? E.g. for DynamoDB it creates the table itself.

Comment (jimbishopp) [reply rfd/0056-sql-backend.md:55]: @r0mant I was concerned users may already have a database with the same name, for whatever reason, and wanted to provide some flexibility. What about making it optional with a default value of "teleport"?

Comment (jimbishopp) [reply rfd/0056-sql-backend.md:55]: I added the default. LMK if you want me to remove it altogether.

File: rfd/0056-sql-backend.md [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+configuration is for self-hosted or cloud-hosted environments.
+
+```yaml
+teleport:
+  storage:
+    # Type of storage backend (postgres or cockroachdb).
+    type: postgres
+
+    # Database connection endpoint.
+    uri: "postgres.example.com:5432"
```
Comment (zmb3) [inline rfd/0056-sql-backend.md:72]: Nit: this is technically not a valid URI. Would `endpoint` be a better name?

**Comment by jimbishopp** [comment]
Updated the RFD based on feedback and added a section about transaction isolation levels.

File: rfd/0056-sql-backend.md (hunk 10 above)
Comment (jimbishopp) [reply rfd/0056-sql-backend.md]: I copied this from the database service section because I wasn't sure if we were going to support a full URI to allow users to set other database configurations as part of the URI. But in the end this doesn't seem necessary.

The other sections of the teleport config use the term `addr` for `host:port` configs, so I went with that to be consistent.

File: rfd/0056-sql-backend.md [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+state: draft
+---
+
+# RFD 56 - SQL Backend
+
+
+## What
+
+A SQL Backend extends Teleport with the ability to store its data in either
+PostgreSQL or CockroachDB with support for self-hosted and cloud-hosted
```
[the remark this answers is not in the document]
Comment (jimbishopp) [reply rfd/0056-sql-backend.md]: I updated this section. LMK if it is clear.

File: rfd/0056-sql-backend.md [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+authors: Jim Bishopp (jim@goteleport.com)
+state: draft
+---
+
+# RFD 56 - SQL Backend
+
+
+## What
+
+A SQL Backend extends Teleport with the ability to store its data in either
```
[the remark this answers is not in the document]
Comment (jimbishopp) [reply rfd/0056-sql-backend.md]: I added a sentence describing what a backend is.

[the remark this answers is not in the document]
Comment (stevenGravy) [reply rfd/0056-sql-backend.md]: I think it would be good to use the language, core cluster state, and if we're providing the other backends.  That language is here https://goteleport.com/docs/setup/reference/backends/

[the remark this answers is not in the document]
Comment (jimbishopp) [reply rfd/0056-sql-backend.md]: @stevenGravy Added the core cluster state term.

**Review by zmb3 — APPROVED** [review]

**Review by r0mant — APPROVED** [review]
